### PR TITLE
Opinionated cleanup of emitRouteError

### PR DESCRIPTION
**Pull request by retrohacker**
## Pre-Submission Checklist

- [x] Opened an issue discussing these changes before opening the PR
- [x] Ran the linter and tests via `make prepush`
- [x] Included comprehensive and convincing tests for changes

## Issues

Closes:

* Issue #1354 

# Changes

* Adding comments
* Simplify edge conditions
* Use `listenerCount` as opposed to creating a copy of the listeners array
* Wait for all event handlers to run before executing cleanup

Alternatively, instead of relying on callbacks, we can assume all event handlers are synchronous and close the request as soon as execution reaches the next line. I believe this would be a significant breaking change. Pushing the requirement onto the user to call a callback on completion seems preferable.

### CODE REVIEW

**Comment by retrohacker** [comment]
It doesn't appear that there is a test covering this behahviour. Do not merge :smile:

File: lib/server.js [hunk 1]
```diff
@@ -208,15 +208,27 @@ function emitRouteError(server, req, res, err) {
...
+    function cleanup() {
         res.send(err);
         server._finishReqResCycle(req, res, null, err);
     }
+
+    var invoked = 0;
+    function cb() {
+        // Check to see if we have received as many callbacks as there are
+        // listeners
+        if (invoked++ === server.listenerCount(name)) {
```
Comment (DonutEspresso) [inline lib/server.js]: I think this is a breaking change, i.e., the response won't get flushed up until all listener callbacks are invoked. I don't have opinions about this opinionated cleanup (badum psh) but probably worth documenting in the migration guide.

**Comment by DonutEspresso** [comment]
With tests, LGTM.

**Comment by retrohacker** [comment]
Closing until this irks me again
